README.md
# express-api-helper

An API response helper module for [Express](http://expressjs.com) apps.

Forked from [Robon Code](https://github.com/roboncode/express-api-helper)

## API

### `ok(req, res, data)`

Respond with `200 OK` and JSON-encoded data.

* `req` express Request
* `res` express Response
* `data` Object

### `badRequest(req, res, errors)`

Respond with `400 Bad Request` and JSON-encoded error object, `{message:String,errors:Array}`.

* `req` express Request
* `res` express Response
* `data` Array (of String) or String

### `unauthorized(req, res)`

Respond with `401 Unauthorized` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `notFound(req, res)`

Respond with `404 Not Found` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `unsupportedAction(req, res)`

Respond with `405 Method Not Allowed` and JSON-encoded error object, `{message:String}`.

* `req` express Request
* `res` express Response

### `invalid(req, res, errors)`

Respond with `422 Unprocessable Entity` and JSON-encoded error object, `{message:String,errors:Array}`.

* `req` express Request
* `res` express Response
* `errors` Array (of String) or String

### `serverError(req, res, error)`

Respond with `500 Internal Server Error` and JSON-encoded error object, `{message:String,error:Object}`.

* `req` express Request
* `res` express Response
* `error` Object

### `requireParams(req, res, params, callback)`

Require that listed parameters are present. Checks for presence of each parameter in `req.body` object if using `express.bodyParser` middleware; otherwise checks for presence of each parameter in `req.params` or `req.query`. If any parameters are missing, invokes `badRequest` with an array of error messages with the form `"Missing required parameter: %s"`.

* `req` express Request
* `res` express Response
* `params` Array (of String) or String
* `callback(err)` Function

## Example

Sample usage:

```javascript
var http = require('http')
  , express = require('express')
  , api = require('express-api-helper')
  , app = express()
  , Post = require('./models/post');

app.configure(function () {
  app.use(express.bodyParser());
  app.use(app.router);
});

app.all('/api/*', function (req, res, next) {
  if (!req.user) return api.unauthorized(req, res);
  next();
});

app.post('/api/posts', function (req, res) {
  api.requireParams(req, res, ['title', 'content', 'authorId'], function (err) {
    if (err) return api.serverError(req, res, err);
    var payload = {
      title: req.body.title,
      content: req.body.content,
      authorId: req.body.authorId
    };
    Post.create(payload, function (err, post) {
      if (err) return api.serverError(req, res, err);
      api.ok(req, res, post.toJSON());
    });
  });
});

app.get('/api/posts', function (req, res) {
  Post.find({}, function (err, posts) {
    if (err) return api.serverError(req, res, err);
    api.ok(req, res, posts.toJSON());
  });
});

app.get('/api/posts/:id', function (req, res) {
  Post.findById(req.params.id, function (err, post) {
    if (err) return api.serverError(req, res, err);
    if (!post) return api.notFound(req, res);
    api.ok(req, res, post.toJSON());
  })
});

http.createServer(app).listen(3000, function () {
  console.log("Express API listening on 3000");
});
```

## Running Tests

To run the tests, clone the repository and install the dev dependencies:

```bash
git clone git://github.com/bryandragon/express-api-helper.git
cd express-api && npm install
make test
```

## License

MIT


CHANGELOG.md
# Change Log

## 0.1.0 (2015-05-02)

  * Added a several statuses
  * Added a friendly message

## 0.0.3 (2014-03-08)

  * When testing for required params in `requireParams()`, use `hasOwnProperty()` and check for params in `req.params` in addition to `req.body` and `req.query`

## 0.0.2 (2014-03-06)

  * Fix bug in `requireParams()` (Issue #1)
  * Upgrade dependency versions

## 0.0.1 (2013-03-10)

  * First version


lib/index.js
'use strict';

var _hasOwnProperty = Object.prototype.hasOwnProperty;

var Status = {
    OK: 200,
    CREATED: 201,
    BAD_REQUEST: 400,
    UNAUTHORIZED: 401,
    FORBIDDEN: 403,
    NOT_FOUND: 404,
    UNSUPPORTED_ACTION: 405,
    TEAPOT: 418,
    VALIDATION_FAILED: 422,
    SERVER_ERROR: 500
};

function statusMessage(status) {
    switch (status) {
        case Status.BAD_REQUEST:
            return 'Bad Request';
        case Status.UNAUTHORIZED:
            return 'Unauthorized';
        case Status.FORBIDDEN:
            return 'Forbidden';
        case Status.NOT_FOUND:
            return 'Not Found';
        case Status.UNSUPPORTED_ACTION:
            return 'Unsupported Action';
        case Status.TEAPOT:
            return 'I am a teapot';
        case Status.VALIDATION_FAILED:
            return 'Validation Failed';
        case Status.SERVER_ERROR:
            return 'Internal Server Error';
    }
}

function jsonResponse(res, body, options) {
    options = options || {};
    options.status = options.status || Status.OK;
    res.status(options.status).json(body);
}

var api = {
    ok: function (data) {
        var body = {
                body: data.body,
                friendlyMessage: data.friendlyMessage || "Object Found"
            };
        jsonResponse(data.res, body, { status: Status.OK });
    },

    created: function (data) {
        var body = {
                body: data.body,
                friendlyMessage: data.friendlyMessage || "Created successfully"
            };
        jsonResponse(data.res, body, { status: Status.CREATED });
    },

    badRequest: function (data) {
        var errors = Array.isArray(data.errors) ? data.errors : [data.errors];

        var body = {
            message: statusMessage(Status.BAD_REQUEST),
            friendlyMessage: data.friendlyMessage || Status.BAD_REQUEST,
            errors: data.errors
        };

        jsonResponse(data.res, body, { status: Status.BAD_REQUEST });
    },

    unauthorized: function (data) {
        var body = { 
            friendlyMessage: data.friendlyMessage || Status.UNAUTHORIZED,
            message: statusMessage(Status.UNAUTHORIZED)
        };

        jsonResponse(data.res, body, { status: Status.UNAUTHORIZED });
    },

    forbidden: function (data) {
        var body = { 
            friendlyMessage: data.friendlyMessage || Status.FORBIDDEN,
            message: statusMessage(Status.FORBIDDEN)
        };

        jsonResponse(data.res, body, { status: Status.FORBIDDEN });
    },

    notFound: function (data) {
        var body = { 
            friendlyMessage: data.friendlyMessage || Status.NOT_FOUND,
            message: statusMessage(Status.NOT_FOUND)
        };

        jsonResponse(data.res, body, { status: Status.NOT_FOUND });
    },

    unsupportedAction: function (data) {
        var body = { 
            friendlyMessage: data.friendlyMessage || Status.UNSUPPORTED_ACTION,
            message: statusMessage(Status.UNSUPPORTED_ACTION)
        };

        jsonResponse(data.res, body, { status: Status.UNSUPPORTED_ACTION });
    },

    teapot: function (data) {
        var body = { 
            friendlyMessage: data.friendlyMessage || Status.TEAPOT,
            message: statusMessage(Status.TEAPOT)
        };

        jsonResponse(data.res, body, { status: Status.TEAPOT });
    },

    invalid: function (data) {
        var errors = Array.isArray(data.errors) ? data.errors : [data.errors];

        var body = {
            message: statusMessage(Status.VALIDATION_FAILED),
            friendlyMessage: data.friendlyMessage || Status.VALIDATION_FAILED,
            errors: data.errors
        };

        jsonResponse(data.res, body, { status: Status.VALIDATION_FAILED });
    },

    serverError: function (data) {
        var error = {};
        if (data.error instanceof Error) {
            error = {
                message: data.error.message,
                stacktrace: data.error.stack
            };
        }
        var body = {
            message: statusMessage(Status.SERVER_ERROR),
            error: error
        };

        jsonResponse(data.res, body, { status: Status.SERVER_ERROR });
    },

    requireParams: function (data, next) {
        var missing = [],
            params = Array.isArray(data.params) ? data.params : [data.params];

        params.forEach(function (param) {
            if (!(data.req.body && _hasOwnProperty.call(data.req.body, param)) && !(data.req.params && _hasOwnProperty.call(data.req.params, param)) && !_hasOwnProperty.call(data.req.query, param)) {
                missing.push('Missing required parameter: ' + param);
            }
        });

        if (missing.length) {
            api.badRequest(data.req, data.res, missing);
        }
        else {
            next();
        }
    }
};

module.exports = api;
